### ISSUE TITLE
Unexpected token '?'

### ISSUE THREAD

**Issue by ballantyne** [fastify/fastify#4262]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.5.2

### Node.js version

18.8

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

11.2

### Description

It seems like it happens when I try to use the process.env.PORT.  I actually don't get to the step where I am calling listen and setting the port, so maybe deep down inside of fastify it is trying to use the default from the environment variable?  Even if I don't interact with the environment variable at all, if it is set, the server crashes.  I am running the script as a child_process that is spawned by a script running as an npm script.  Not sure any that would make any difference.  I don't know if I have ever seen '??' before.  Is '??' an operator?

It seems like a strange error.  I don't even have to call listen and it fails, it happens when I require the module, even before invoking it.

var Fastify = require('fastify'); // <= this is when it fails.
var fastify = Fastify({}) 

If I don't set the environment variable, it works.  If I don't set the variable does the script inherit the env from the parent process?  Maybe that could be the reason?  Not really sure what the problem is.

```
/home/ballantyne/projects/project_name/node_modules/fastify/lib/server.js:38
      const hostPath = listenOptions.path ? [listenOptions.path] : [listenOptions.port ?? 0, listenOptions.host ?? 'localhost']
                                                                                        ^

SyntaxError: Unexpected token '?'
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/home/ballantyne/projects/project_name/node_modules/fastify/fastify.js:34:54)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
```

### Steps to Reproduce

// server.js
```
var Fastify = require('fastify') // <= fails here

fastify = Fastify({});
fastify.get('/', function (request:any, reply:any) {
  reply.send("hello")
})

fastify.listen({ port: 5555, host: 'localhost' }, function (err:any, address:any) {
  if (err) {
    fastify.log.error(err)
    process.exit(1)
  }
})
```

// script that spawns server
```
var server = spawn('node', ['script/with/fastify/example/from/readme/server.js'], {cwd: /absolute/path/to/enclosing/folder, stdio: 'inherit', env: {PORT: 5555 }});
```

### Expected Behavior

The script should start listening to port 5555.

**Comment by jsumners** [comment]
Double check the version of Node your system is using to start the Fastify based application. It should be at least version 14.

**Comment by ballantyne** [comment]
$ node -v
v18.8.0

The version of fastify I have installed is actually 4.5.3, but I also
noticed the problem with 4.4.0

**Comment by jsumners** [comment]
That's certainly not the version that is launching your application according to the error you have posted. The coalescing operator has been supported since Node 14.5.0. See https://node.green/#ES2020-features--nullish-coalescing-operator-----

**Comment by ballantyne** [comment]
I just did a console.log(process.execPath) and it prints "/usr/local/bin/node"

and when I ask for the version I get this:
```
$ /usr/local/bin/node -v
v18.8.0
```

I downloaded the source of node from:
https://nodejs.org/dist/v18.8.0/node-v18.8.0.tar.gz

**Comment by mcollina** [comment]
Try this:

1. copy the example in the readme, save it as `example.js`
2. run it with `node example.js`

We expect this to work.

---

What James is saying is that for some reason in your chain of scripts/npm/etc, you are picking up an old node version to run Fastify.

What operating system are you in?

**Comment by ballantyne** [comment]
I put the console.log(process.execPath) in both of the files that get
called.  Both showed the same result.  The npm user agent also indicated
18.8.0 as the node version.  I use debian.

if you are not interested in this issue i am not going to hassle you.  Just
thought i would let you know.  I wasn’t sure what your fancy ‘??’ did so i
thought i would inquire.

**Comment by Eomm** [comment]
I would check the whole `process` object.

[`spawn`](https://nodejs.org/api/child_process.html#child_processspawncommand-args-options) has a `env` option: you can force it. I think that the new spawn process is not running the same `node` installation as when you run `node example.js`

**Comment by Uzlopak** [comment]
what does `console.log(process.version)` say?

**Comment by ballantyne** [comment]
When I pass the env variable, process.env just becomes {PORT: 5555} and it
doesn't have any other variables and it fails the way I was saying:
```
{ PORT: '5555' }
/home/ballantyne/projects/movement/benchmark/fastify/node_modules/fastify/lib/server.js:38
      const hostPath = listenOptions.path ? [listenOptions.path] :
[listenOptions.port ?? 0, listenOptions.host ?? 'localhost']

            ^

SyntaxError: Unexpected token '?'
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js
(internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous>
(/home/ballantyne/projects/movement/benchmark/fastify/node_modules/fastify/fastify.js:34:54)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js
(internal/modules/cjs/loader.js:1027:10)
```

If I don't pass the env to the spawn command it inherits the parents env
which has
```
{
  NODE: '/usr/local/bin/node',
  npm_config_globalconfig: '/usr/local/etc/npmrc',
  npm_config_init_module: '/home/ballantyne/.npm-init.js',
  npm_config_global_prefix: '/usr/local',
  npm_config_node_gyp:
'/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js',
  npm_config_cache: '/home/ballantyne/.npm',
  npm_config_prefix: '/usr/local',
  npm_config_noproxy: '',
  npm_node_execpath: '/usr/local/bin/node',
  npm_config_user_agent: 'npm/8.18.0 node/v18.8.0 linux x64
workspaces/false',
  npm_config_metrics_registry: 'https://registry.npmjs.org/'
}
// etc
```

**Comment by ballantyne** [comment]
I sent this as an email but it didn't show up:

When I try console.log(process.version) "v18.8.0" shows in the script that calls spawn and "v18.8.0" shows in the script that has the code that is from the readme.

**Comment by ballantyne** [comment]
I think that I figured out sort of what was happening.  When the only env variable is {PORT: 5555} it is finding the node that was installed by apt at /usr/bin/node rather than the node that was installed from source at /usr/local/bin/node.  

So it was version 12.22.12.  James was correct about the version being old.  Strange that it showed the newer version inside both scripts when I printed it to console.log.  /usr/local/bin is before /usr/bin in my path, so since it doesn't have an a path in the process.env how does it choose?  Does node choose or does debian choose?  How does node choose if it chooses and has no path?  I honestly don't know the reason why it behaved the way that it did.  I think I have the code I am working on working the way I need now.  Thanks for your response.

**Comment by ballantyne** [comment]
I did recompile from the source last night.  Maybe that is the reason I have a different result?  Probably not.  I ran a script that I wrote to install node so I downloaded the code from the same place.  Not really sure why there would be a different between yesterday and today.  Today, the 12.22.12 version is showing up in the console.log(process.version) script that is spawned if I don't specific the exact path for node.

I just took another look again.  If I pass an env of {PORT: 5555} it uses the /usr/bin/node which is 12.22.12, but if I don't pass in any environment variables and the spawn command inherits from the parent it uses /usr/local/bin/node which is 18.8.0.  

So, if I don't specify the absolute path of the binary I am trying to use using process.execPath and I make it so that the only environment variable that is passed into the spawned process is {PORT: 5555}, the spawned process chooses /usr/bin/node even though /usr/local/bin/node occurs earlier in my path.

I realize this seems like it is related to the way that either debian or node works rather than fastify.  The fastify issue is that it uses syntax that is too new for the old version of node that I have installed at /usr/bin/node.

Anyway, thought I would try to give as much information about the issue so that it makes sense.
